tests/InfillTest.cpp

```diff
@@ -0,0 +1,265 @@
+        for (int i_method = 0; i_method < static_cast<int>(EFillMethod::NONE); ++i_method)
+        {
+            const EFillMethod method = static_cast<EFillMethod>(i_method);
+            if (std::find(skip_methods.begin(), skip_methods.end(), method) == skip_methods.end()) // Only use if not in skipped.
+            {
+                methods.push_back(method);
+            }
+        }
+
+        std::vector<coord_t> line_distances = { 400, 600, 800, 1200 }; // TODO?: Gyroid fails the 'fill less than 100% of area available' with values close to the line width, like 350.
+
+        std::vector<InfillTestParameters> parameters_list;
+        size_t test_polygon_id = 0;
+        for (const Polygons& polygons : shapes)
+        {
+            for (const EFillMethod& method : methods)
+            {
+                for (const coord_t& line_distance : line_distances)
+                {
+                    parameters_list.push_back(generateInfillToTest(InfillParameters(method, false, false, line_distance), test_polygon_id, polygons));
+                    parameters_list.push_back(generateInfillToTest(InfillParameters(method, false, true, line_distance), test_polygon_id, polygons));
+                    parameters_list.push_back(generateInfillToTest(InfillParameters(method, true, false, line_distance), test_polygon_id, polygons));
+                    parameters_list.push_back(generateInfillToTest(InfillParameters(method, true, true, line_distance), test_polygon_id, polygons));
+                }
+            }
+            ++test_polygon_id;
+        }
+
+        return parameters_list;
+    }
+
+    class InfillTest : public testing::TestWithParam<InfillTestParameters> {};
+
+    INSTANTIATE_TEST_CASE_P(InfillTestcases, InfillTest, testing::ValuesIn(generateInfillTests()), [](testing::TestParamInfo<InfillTestParameters> info) { return info.param.name; });
+
+    TEST_P(InfillTest, TestInfillSanity)
+    {
+        InfillTestParameters params = GetParam();
+        ASSERT_TRUE(params.valid) << params.fail_reason;
+        ASSERT_FALSE(params.result_polygons.empty() && params.result_lines.empty()) << "Infill should have been generated.";
+
+        const double available_area = std::abs(params.outline_polygons.area());
+        const double expected_infill_area = (available_area * infill_line_width) / params.params.line_distance;
+        const double total_infill_area = (params.result_polygons.polygonLength() + params.result_lines.polyLineLength()) * infill_line_width / getPatternMultiplier(params.params.pattern);
+        ASSERT_GT((coord_t)available_area, (coord_t)total_infill_area) << "Infill area should allways be less than the total area available.";
+        ASSERT_NEAR((coord_t)total_infill_area, (coord_t)expected_infill_area, (coord_t)(total_infill_area * 0.15)) << "Infill area should be within 15% of expected size."; // TODO: 10 ~ 20% is actually quite bad?
```

Comment:
Not something that can be automated really, but did you see the test case (export to SVG using the `SVG.h` debugger class) and resulting infill and why it's so bad? If the result is correct then the current assertion is fine. But otherwise there may be a bug there.

Reply:
I did, and it showed me that the inner hole polygons in my test cases had their vertex order the wrong way around (making them not holes).

I also fixed up this code a bit taking into account the discrete nature of infill.